std/exception.d

```diff
@@ -584,8 +597,11 @@ T errnoEnforce(T, string file = __FILE__, size_t line = __LINE__)
     return value;
 }
 
-
+// @@@DEPRECATED_2019-07@@@
 /++
+    $(RED Deprecated. Please use $(LREF enforce) instead. This function will be removed
+               in July 2019.)
```

Comment:
Hmm, are we still doing deprecations by dates? I thought there was consensus that deprecating by D versions is overall better.

Reply:
Ehm our deprecation process definitely needs some guidelines because I wasn't aware of this.
I would prefer D versions too.

Reply:
By version would be nice. @MartinNowak what do you think?

Reply:
I've been using month/year, because historically, our releases haven't been consistent enough to use any kind of release numbers when deprecating things. I can easily pick a month a year from now and put that in the message. Historically, that hasn't worked at all with versions. Martin has improved that situation considerably, but they're still not completely consistent. And as a user, it's far easier to look at a date and see about when the symbol is going away than it is to try and figure out when on earth a specific version is going to be released. If the date is a year from now, then it's clear that I have a year. If it's a version number, then it's not at all clear how long I have. I'd have to examine what the release schedule has been and estimate about when the symbol is going away.

Reply:
However, the date in this case should be more like 2019-02, since it's supposed to be about a year from now (after which, it will be undocumented and have a non-ddoc comment indicating a date about a year from then for when it will be fully removed).

Reply:
> And as a user, it's far easier to look at a date and see about when the symbol is going away than it is to try and figure out when on earth a specific version is going to be released. 

Err, sorry, I think that's a bit removed from reality. Essentially that's only true (and useful) if you're always using the latest version of DMD, and use nothing but the latest version of DMD. While this is mostly true for us who contribute to DMD and core D components frequently, it likely isn't for most D users.

Using dates is less useful for answering these questions:
- Upgrading to which version of DMD will break my code?
- Downgrading to which version of DMD will get my code to build again?
- In which version of LDC or GDC will this code stop working?
- Will upgrading to the next release of my Linux distribution mean I'll have fix my code?

Essentially, dates become a lot less useful in any situation where the user is not using the latest version of Phobos, whether that's thanks to a rolling release Linux distro, using D-apt, or religious frequent manual updates. In all other cases, I believe version numbers would be more useful and easier to look up.

Reply:
Well, technically, it keeps working for about a year after it's removed from the docs, so I'd be inclined to argue that the imprecision with regards to how it matches up with ldc releases or whatnot doesn't matter much, and the site always has the latest docs only, but if we get to the point that we have the docs for previous releases like we've discussed before, then dates definitely become worse.

So, fine, if you guys really want version numbers, then we go with version numbers, but it's going to make the actual time frame of deprecations more inconsistent, since it'll be harder to target a year ahead, but while they've usually been in the ballpark of where they should be, they don't always get moved along quite when they should anyway.

In that case, let's change it to say

    // @@@DEPRECATED_2.084@@@

and

    This function will be removed in 2.084.

and I'll have to create some PRs to fix the existing deprecations.

Reply:
> since it'll be harder to target a year ahead,

I don't think we need to try to target deprecations using calendar time, just use "N major versions ahead" instead of "N months ahead".

> but while they've usually been in the ballpark of where they should be, they don't always get moved along quite when they should anyway.

Yeah, one advantage of using dates is that when we don't honor our own deprecation deadlines, it's easier to excuse that for dates than version numbers.

Reply:
> And as a user, it's far easier to look at a date and see about when the symbol is going away than it is to try and figure out when on earth a specific version is going to be released.

Thinking more on this, I think I understand better what you meant here.

If you are maintaining an open-source project (especially a library), a deprecation date (on a symbol used in your code) will tell you what's the latest point in time that the deprecation is guaranteed to still work for everyone, everywhere, so you don't need to really worry about it until then. Anything past that date, though, and there appears the chance that someone, somewhere is going to start having trouble building your software.

So... I'd say version numbers are more useful if you're writing code for yourself or your company, and deprecation dates are more useful if you're writing open-source software.

Reply:
> In that case, let's change it to say

OK. Changed. We should probably add a short documentation in the wiki or as a markdown document this repo about the deprecation process.